util/slice.cc

```diff
@@ -209,4 +209,33 @@ const SliceTransform* NewNoopTransform() {
   return new NoopTransform;
 }
 
+PinnableSlice::PinnableSlice(PinnableSlice&& other) {
+  *this = std::move(other);
+}
+
+PinnableSlice& PinnableSlice::operator=(PinnableSlice&& other) {
+  if (this != &other) {
+    Cleanable::Reset();
+    Cleanable::operator=(std::move(other));
+    size_ = other.size_;
+    pinned_ = other.pinned_;
+    if (pinned_) {
+      data_ = other.data_;
+      // When it's pinned, buf should no longer be of use.
+    } else {
+      if (other.buf_ == &other.self_space_) {
+        self_space_ = std::move(other.self_space_);
+        buf_ = &self_space_;
+        data_ = buf_->data();
+      } else {
+        buf_ = other.buf_;
+        data_ = other.data_;
+      }
+    }
+    other.size_ = 0;
```

Comment:
Would be a good idea to also reset ```other.buf_``` to its ```self_space_```. What would be the value of ```other.data_```?

Reply:
to be consistent, after resetting `other.buf_` to `other.self_space_`, `other.data_` needs to be set to `other.buf_`.